### PR TITLE
feat(Design-System/Stack): Creating layout component Stack

### PR DESCRIPTION
**Pull request by Benrajalu**
**What is the problem this PR is trying to solve?**
Even when using shared components, developers still have to write CSS to place and position these components in their apps. This is repetitive work with no real added value, and it's where devs may use custom spacing values instead of design tokens.

**What is the chosen solution to this problem?**
`Stack` is a layout component that handles what `display: flex` does. It comes with opinionated props for spacing (padding and gap between children) that are limited to design tokens. 

**Please check if the PR fulfills these requirements**

- [ ] The PR have used `yarn changeset` to a request a release from the CI if wanted.
- [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features) And [non reg](./screenshots.md) done before need review
- [x] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: packages/design-system/src/components/Stack/Stack.stories.mdx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+            <Stack
+                direction="vertical"
+                padding="S"
+                gap="XS"
+                align="start"
+                justify="start"
+                as="ul"
+                height={tokens['coral-size-limit']}
+                width={tokens['coral-size-limit']}
```
Comment (frassinier) [inline packages/design-system/src/components/Stack/Stack.stories.mdx:55]:
```suggestion
                height={tokens.coralSizeLimit}
                width={tokens.coralSizeLimit}
```

File: packages/design-system/src/components/Stack/Stack.stories.mdx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                gap="S"
+                align="start"
+                justify="start"
+            >
+                <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.</p>
+                <Button.Primary onClick={clickHandler}>Click here</Button.Primary>
+            </Stack>
+
+            <Stack
+                direction="vertical"
```
Comment (frassinier) [inline packages/design-system/src/components/Stack/Stack.stories.mdx:48]: if the goal is to get a 1:1 mapping with CSS one then
```suggestion
                direction="row"
```

File: packages/design-system/package.json [hunk 3]
```diff
@@ -66,7 +67,7 @@
     "@testing-library/react": "^12.1.2",
     "@testing-library/user-event": "^13.3.0",
     "@types/classnames": "^2.2.11",
-    "@types/react": "^16.14.2",
+    "@types/react": "^17.0.37",
```
Comment (frassinier) [inline packages/design-system/package.json:70]: Oh?
```suggestion
    "@types/react": "^17.0.37",
```

**Review by frassinier — APPROVED** [review]

**Comment by frassinier** [comment]
Please hit `yarn changeset` at root folder to fill up the changelog and be able to trigger a release

File: packages/design-system/package.json (hunk 3 above)
Comment (Benrajalu) [reply packages/design-system/package.json:70]: My bad

File: packages/design-system/yarn.lock [hunk 4]
```diff
@@ -0,0 +1 @@
+../../yarn.lock
```
Comment (romainseb) [inline packages/design-system/yarn.lock:1]: Why do we need the lockfile here where it does not seems to be needed in the other packages in the monorepo ?

File: packages/design-system/tsconfig.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
-  "extends": "../../node_modules/@talend/scripts-config-typescript/tsconfig.json",
-  "include": ["cypress/*", "src/*", "custom.d.ts"],
+  "extends": "@talend/scripts-config-typescript/tsconfig.json",
```
Comment (romainseb) [inline packages/design-system/tsconfig.json:2]: Should this dep be in dev dep in the package so ?

File: packages/design-system/yarn.lock (hunk 4 above)
Comment (Benrajalu) [reply packages/design-system/yarn.lock:1]: Good question, TS really wanted it IIRC, @frassinier?

File: packages/design-system/tsconfig.json (hunk 5 above)
Comment (Benrajalu) [reply packages/design-system/tsconfig.json:2]: I believe you have it as a dep of the monorepo?

File: packages/design-system/yarn.lock (hunk 4 above)
Comment (frassinier) [reply packages/design-system/yarn.lock:1]: It's for Security Scan purpose

File: packages/design-system/src/components/Stack/Stack.stories.mdx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+<StackVertical gap="XS" as="ul">
+    <li>List entry</li>
+    <StackItem as="li" align="center" grow>
+        List entry 2
+    </StackItem>
+    <li>List entry 3</li>
+</StackVertical>
+```
+
+## Usage
```
Comment (jsomsanith-tlnd) [inline packages/design-system/src/components/Stack/Stack.stories.mdx:47]: We need doc for the 3 spacing mode (size vs {x,y} vs {top,bottom,left, right})

For padding or margin on 1 side, you have to specify all values, even the ones you want to leave at 0. The code base the shape to `top` so in reality, it is mandatory to set the top value to have the other sides taken into account.
We should have more precision in the doc to tell user to set all the values in this mode.

File: packages/design-system/src/components/Stack/StackItem.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	overflow?: keyof typeof overflowOptions;
+	as?: typeof possibleAsTypes[number];
+};
+
+export const StackItem = forwardRef(function StackItem(
+	{
+		as = 'div',
+		children,
+		grow = false,
+		shrink = true,
```
Comment (jsomsanith-tlnd) [inline packages/design-system/src/components/Stack/StackItem.tsx:37]: Why is it true ?
In the css, the default item shrink is 0. But in the code you force it a 1 except if the user force it at 0.

Comment (Benrajalu) [reply packages/design-system/src/components/Stack/StackItem.tsx:37]: Initial flex-shrink is 1 https://developer.mozilla.org/en-US/docs/Web/CSS/flex-shrink

File: packages/design-system/src/components/Stack/Stack.stories.mdx (hunk 6 above)
Comment (Benrajalu) [reply packages/design-system/src/components/Stack/Stack.stories.mdx:47]: Sure, well spotted.

File: packages/design-system/src/components/Stack/StackItem.tsx (hunk 7 above)
Comment (jsomsanith-tlnd) [reply packages/design-system/src/components/Stack/StackItem.tsx:37]: ok 👍

File: packages/design-system/src/components/Stack/Primitive/StackPrimitive.tsx [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+			];
+		}
+
+		return [];
+	}
+
+	function getAlignContent() {
+		if ('alignContent' in props) {
+			const { alignContent, ...rest } = props;
+			spreadableProps = rest;
```
[the remark this answers is not in the document]
Comment (Benrajalu) [reply packages/design-system/src/components/Stack/Primitive/StackPrimitive.tsx:190]: Not sure I follow. I can't figure out how I can `const` this while keeping the conditional spreading depending on whether alignContent exists...

File: packages/design-system/src/components/Stack/Stack.stories.mdx (hunk 6 above)
Comment (Benrajalu) [reply packages/design-system/src/components/Stack/Stack.stories.mdx:50]: Done!

File: packages/design-system/src/components/Stack/Primitive/StackPrimitive.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (Benrajalu) [reply packages/design-system/src/components/Stack/Primitive/StackPrimitive.tsx]: My bad, fixed

**Review by jsomsanith-tlnd — APPROVED** [review]

**Pull request merged**